driver-bitfury16.c

```diff
@@ -0,0 +1,4589 @@
+					else {
+						renonce_chip_address[board_id].board_id  = -1;
+						renonce_chip_address[board_id].bcm250_id = -1;
+						renonce_chip_address[board_id].chip_id   = -1;
+					}
+				} else
+					renonce_chips++;
+			} else {
+				renonce_chip_address[board_id].board_id  = -1;
+				renonce_chip_address[board_id].bcm250_id = -1;
+				renonce_chip_address[board_id].chip_id   = -1;
+			}
+		}
+
+		if (renonce_chips != info->renonce_chips) {
+			applog(LOG_ERR, "%s: expected to find [%d] renonce chips, but found only [%d]",
+					bitfury->drv->name, opt_bf16_renonce, renonce_chips);
+
+			info->renonce_chips = renonce_chips;
+		}
+	} else {
+		for (board_id = 0; board_id < CHIPBOARD_NUM; board_id++) {
+			renonce_chip_address[board_id].board_id  = -1;
+			renonce_chip_address[board_id].bcm250_id = -1;
+			renonce_chip_address[board_id].chip_id   = -1;
+		}
+	}
+
+#ifdef FILELOG
+	info->logfile = fopen(LOGFILE, "a");
+	if (info->logfile == NULL)
+		applog(LOG_ERR, "%s: failed to open logfile [%s]: %s",
+				bitfury->drv->name, LOGFILE, strerror(errno));
+#endif
+
+	/* exit if no boards present */
+	if ((info->chipboard_num == 0) ||
+		(info->active_chipboard_num == 0)) {
+		deinit_x5(bitfury);
+
+		/* close devices */
+		close_spi_device(SPI_CHANNEL1);
+		close_spi_device(SPI_CHANNEL2);
+		close_ctrl_device();
+		close_uart_device(UART_CHANNEL1);
+		close_uart_device(UART_CHANNEL2);
+
+#ifdef FILELOG
+		fclose(info->logfile);
+#endif
+
+		applog(LOG_ERR, "%s: no boards present. exiting...",
+				bitfury->drv->name);
+
+		free(info);
+		free(bitfury);
+
+		return;
+	}
+
+	mutex_init(&info->nonces_good_lock);
+
+	if (!add_cgpu(bitfury))
+		quit(1, "%s: %s() failed to add_cgpu",
+				bitfury->drv->name, __func__);
+
+	info->initialised = true;
+
+	applog(LOG_INFO, "%s: chip driver initialized", bitfury->drv->name);
+#ifdef FILELOG
+	filelog(info, "%s: cgminer started", bitfury->drv->name);
+#endif
+}
+
+static uint8_t chip_task_update(struct cgpu_info *bitfury, bf_chip_address_t chip_address)
+{
+	uint8_t i;
+	uint8_t ret = 0;
+	bf_works_t work;
+	time_t curr_time_t;
+	struct timeval curr_time;
+
+	uint8_t board_id  = chip_address.board_id;
+	uint8_t bcm250_id = chip_address.bcm250_id;
+	uint8_t chip_id   = chip_address.chip_id;
+
+	struct bitfury_info *info = (struct bitfury_info *)(bitfury->device_data);
+	bf_cmd_buffer_t* cmd_buffer = &info->chipboard[board_id].cmd_buffer;
+
+	if (cmd_buffer->status != EMPTY)
+		return -1;
+
+	switch (info->chipboard[board_id].bcm250[bcm250_id].chips[chip_id].status) {
+
+	/* fill chip buffer with toggle task */
+	case UNINITIALIZED:
+		applog(LOG_DEBUG, "%s: chipworker_thr: chip [%d:%d:%2d], prepare toggle_cmd",
+				bitfury->drv->name,
+				board_id, bcm250_id, chip_id);
+
+		uint8_t toggle[4] = { 0xa5, 0x00, 0x00, 0x02 };
+
+		ret = cmd_buffer_push(cmd_buffer,
+				info->chipboard[board_id].bcm250[bcm250_id].channel_depth, chip_address, chip_address,
+				work, 0, CHIP_CMD_TOGGLE, 3, toggle);
+
+		if (ret < 0)
```

Comment:
driver-bitfury16.c:1495:11: warning: comparison is always false due to limited range of data type [-Wtype-limits]
   if (ret < 0)